### PR TITLE
[slack] Give gist link for detected SELECT statements

### PR DESCRIPTION
**Pull request by Harshg999**
## What changes were proposed in this pull request?

- Removed ephemeral messages -- Cant be used in unfurling
- Detect SELECT statements
- Check Slack user permission to create gist
- Create gist and send link
- Can be shown in the message thread too if we don't want to clutter up the Slack channels

## How was this patch tested?

Screenshot 2021-05-17 at 5 06 43 PM

### CODE REVIEW

File: desktop/core/src/desktop/lib/botserver/views.py [hunk 1]
```diff
@@ -96,22 +96,33 @@ def handle_on_message(channel_id, bot_id, elements, user_id):
...
-
+      detect_select_statement(host_domain, is_http_secure, channel_id, user_id, text.lower())
+
+
+def detect_select_statement(host_domain, is_http_secure, channel_id, user_id, statement):
+  slack_user = check_slack_user_permission(host_domain, user_id)
+  user = get_user(channel_id, slack_user)
+
+  gist_doc = Document2.objects.create(
+    name='Mysql Query blah',
```
Comment (romainr) [inline desktop/core/src/desktop/lib/botserver/views.py:107]: blah --> ''

File: desktop/core/src/desktop/lib/botserver/views.py [hunk 2]
```diff
@@ -96,22 +96,33 @@ def handle_on_message(channel_id, bot_id, elements, user_id):
...
+    type='gist',
+    owner=user,
+    data=json.dumps({'statement_raw': statement}),
+    extra='mysql',
```
Comment (romainr) [inline desktop/core/src/desktop/lib/botserver/views.py:111]: As we don't know the dialect we should default to the default one (if not we lose a lot in usability).

They are ways to detect which type of SQL it is but bit more complex and to see later.

File: desktop/core/src/desktop/lib/botserver/views.py [hunk 3]
```diff
@@ -96,22 +96,33 @@ def handle_on_message(channel_id, bot_id, elements, user_id):
...
+    parent_directory=Document2.objects.get_gist_directory(user)
+  )
+
+  gist_link = '%(scheme)s://%(host)s/hue/gist?uuid=%(uuid)s' % {
+    'scheme': 'https' if is_http_secure else 'http',
+    'host': host_domain,
+    'uuid': gist_doc.uuid,
+  }
```
Comment (romainr) [inline desktop/core/src/desktop/lib/botserver/views.py:119]: Isn't this a duplication of the gist creation functionality?

**Review by romainr — COMMENTED** [review]
Nice experience!

File: desktop/core/src/desktop/lib/botserver/views.py (hunk 3 above)
Comment (Harshg999) [reply desktop/core/src/desktop/lib/botserver/views.py:119]: Yes, however gist_create is extracting required params (name, statement, etc) from the request argument @romainr
Also, to create link it takes host and scheme from request

Is there a need to create a helper function for gist_create  or am I missing something @romainr ??

Comment (romainr) [reply desktop/core/src/desktop/lib/botserver/views.py:121]: Yes missing that we are duplicating create gist functionality?

Comment (Harshg999) [reply desktop/core/src/desktop/lib/botserver/views.py]: Created a helper function for gist creation to remove duplication

File: desktop/core/src/desktop/lib/botserver/views.py [hunk 4]
```diff
@@ -100,22 +101,22 @@ def handle_on_message(channel_id, bot_id, elements, user_id):
       _send_message(channel_id, bot_message)
 
     if text.lower().startswith('select'):
-      detect_select_statement(channel_id, user_id, text.lower())
+      detect_select_statement(host_domain, is_http_secure, channel_id, user_id, text.lower())
```
Comment (romainr) [inline desktop/core/src/desktop/lib/botserver/views.py:104]: `detect_select_statement` --> `_handle_select_statement`

File: desktop/core/src/desktop/api2.py [hunk 5]
```diff
@@ -924,37 +924,45 @@ def gist_create(request):
...
   description = request.POST.get('description', '')
 
-  if not name:
-    name = _('%s Query') % gist_type.capitalize()
+  response = _gist_create(request.get_host(), request.is_secure(), request.user, statement, gist_type, name)
+
+  return JsonResponse(response)
+
+
+def _gist_create(host_domain, is_http_secure, user, statement, gist_type, name=''):
```
Comment (romainr) [inline desktop/core/src/desktop/api2.py:938]: Nice!

This now makes it trivial to test and share within the app!

**Review by romainr — APPROVED** [review]
Nice!

File: desktop/core/src/desktop/api2.py (hunk 5 above)
Comment (Harshg999) [reply desktop/core/src/desktop/api2.py:938]: Agreed!

**Pull request merged**